dawn/docs/parallel_model.rst

```diff
@@ -0,0 +1,251 @@
+
+Formal Definition of the Dawn Parallel Model
+============================================
+
+Contract to the user from HIR
+-----------------------------
+
+Dawn will generate code based on the operations passed from high-level
+intermediate representation. It's specification can be found `here`_.
+
+Dependencies
+~~~~~~~~~~~~
+
+Generally, stencil computation in dawn differentiates between horizontal
+and the vertical dimension. The assumption is that individual statements
+of a stencil are embarrassingly parallel in the horizontal. For that
+reason, there is no guarantee that any horizontal point's computation
+will be finished before any other in each user statement.
+
+Vertical Dependencies
+^^^^^^^^^^^^^^^^^^^^^
+
+Each vertical region of a stencil specifies a loop order of execution.
+Dawn will uphold the following rules:
+
+-  If the loop order is forward, any statement with a vertical
+   dependency on other statements executed on a vertical level with
+   index ``k`` will be executed after each statement that it depends on
+   that appears previously in the current vertical region or in any
+   previous vertical region was executed on every vertical level with
+   indices ``k-1`` and smaller.
+-  If the loop order is backward, any statement with a vertical
+   dependency on another statement executed on a vertical level with
+   index ``k`` will be executed after each statement that it depends on
+   that appears previously in the current vertical region or in any
+   previous vertical region was executed on every vertical level with
+   indices ``k+1`` and bigger.
+
+An example of this would be the following
+
+.. code:: c++
+
+   1 vertical_region(k_start, k_end)
+   2   one = 1
+   3 vertical_region(k_start, k_end)
+   4   two = one[k - 1]
+
+Here, for every level with index ``k``, line two will be executed before
+line 4.
+
+Horizontal Dependencies
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Irrespective of the loop order every statement with a horizontal
+dependency on other statements will be executed after all statements it
+depends on are executed on the full horizontal plane in the same
+vertical level.
+
+An example of this is the following
+
+.. code:: c++
+
+   1 vertical_region(k_start, k_end)
+   2   one = 1
+   3   two = one[i - 1]
+
+For every vertical level, we guarantee that line two will be executed on
```

Comment:
Not sure if this is true. You say here that we allow vertical-diagonal dependencies:
```
   1 vertical_region(k_start, k_start)
   2   one = 1
   3 vertical_region(k_start + 1, k_end)
   4   one = 1
   5   two = one[i - 1, k - 1]
```
Is this a real pattern we should support? This are not "pure vertical solvers" anymore.

Reply:
That is a pattern we never encountered in all of comso. Unclear if we need to support it

Reply:
Yes, I know. This is why I bring this point (it was already brought up many times before). This simplifies a lot when thinking about the problems and we can fully split between horizontal stencils and vertical solvers.

Reply:
I agree that we never encounter it. Might it appear indirectly, because we do something like
```
vertical_region(k_start, k_end)
    zero = 0
    one = zero[k - 1]
    two = one[i + 1]
```
I think the consequence is the same: At some point we need to split into two multistages (because two accesses `zero[k - 1, i + 1]`. 

If I am right and this can also happen indirectly, it might be better to support the pattern, because it's too hard for the user to know when he uses this pattern. But it is not yet fully thought through.

Reply:
yes, that is another interesting example. I think this is a real use case. 
What about this in the VerticalAdvectionFunctions ? 
`Average<Direction, DirectionFactor, NestedShift<Z, 1>>(eval, wcon());`
As far as we have staggering, I think is normal to do stencils (averages, etc) in the i,j dimension for a displaced position in k. But I think all these cases are stencils and not solvers. I think solvers should be always accessing a column. It looks weird that a solver is accessing an intermediate value of the diagonals of the solve of a neighbour column, or at least I dont know what that means. Therefore I would say that we need to support this, directly and/or indirectly (which as pointed out by lukas does not make any difference). So I guess adding the requirement that solvers are purely vertical to any of the semantic options would do the job. 
However @lukasm91 added the example that would indirectly violate this condition. If I can also reformulate the example: 
```
var one
vertical_region(k_start, k_start)
    zero = 0
vertical_region(k_start+1, k_end)
    zero = zero[k - 1]
    one = zero
    two = one[i + 1]
```
So what we could do is to add a check (or a pass) that for any vertical region, any horizonal stencil access is not accessing and field that has have a vertical solver access. This is probably not trivial since we need to traverse backwards the DAG of dependencies. 
Additionally, we could have the following: 
```
var one
vertical_region(k_end, k_end)
    zero = 0
vertical_region(k_end-1, k_start+1)
    zero = zero[k + 1]
vertical_region(k_start, k_start)
    zero = zero[k + 1]
    one= zero[i + 1]
```
that means we have a forward (not in the example)/backward solver, and at the end of the solver for level k_start we still have a solver access (`zero=zero[k+1]` ). After that stmt, actually the solution is in place in the fields and we could access `one = zero[i+1]`. When the solver is finished is up to the user, it could happen at a boundary layer (k_end, k_start) or at any other level. 
So 
* option 1: we dont add any protection/check and let the responsibility to the user.
* option 2: we do check, and force the user to close the vertical_region and open a new one before he can access the solutions of a solver. However, that is still complicated, because as output of some optimizations, the example I made before where we access the output of a solver at the last level should be legal. So even if we force the user to split vertical_regions, we can not check this state internally anymore after some optimizations happen.

Reply:
Option 3: We just generate the right code. This might lead to very slow codes (for CUDA, a kernel call in each k-iteration), but at least we have legal code, and we can diagnose to the user why his code is small.